### PR TITLE
Port to Acorn 6 plugin API

### PR DESCRIPTION
**Pull request by marijnh**
Note that Acorn 6 isn't out yet, but before I finalize the new plugin API I'm taking a stab at porting popular plugins to make sure it causes no problems. I'd like your feedback on both my changes here and the new interface (you can read about it in the [readme file](https://github.com/acornjs/acorn/blob/master/README.md#plugins) on the current master branch of Acorn).

The main change, apart from mechanical porting, is that there's no longer a concept of injecting. Importing the plugin gives you a plugin value that you pass to `acorn.Parser.extend` to get an extended parser. This is a little less magical and easier to compose, but it does significantly change the package's interface.

What do you think? I'm planning to release Acorn 6.0.0 this week, at which point, if you are happy with the changes, it'd make sense to merge and release this.

### CODE REVIEW

**Comment by marijnh** [comment]
I forgot to mention, I also fixed a bug where `import ("x")`, with a space before the paren, failed to parse in the old code, because it only looked at the character directly after the import token.

**Comment by marijnh** [comment]
Acorn 6.0.0 has been released now.

**Comment by adrianheine** [comment]
This only works for me if babel doesn't transpile classes. I used the following `.babelrc`:
```
{
  "plugins": ["transform-es2015-modules-commonjs"]
}
```

**Comment by marijnh** [comment]
Oh, I only ran the tests, which appear to use the untransformed source. I'll take a look.

**Comment by marijnh** [comment]
What goes wrong? If I make the tests load the file from `lib/` they still pass.

**Comment by adrianheine** [comment]
For me, babel transpiles the class and tries to call the constructor of a super class without `new`.

**Comment by marijnh** [comment]
Isn't the superclass also compiled down to ES5?

**Comment by adrianheine** [comment]
Huh, I cannot reproduce the issue and I can't think of a situation where the untranspiled `Parser` class would appear in my setup.

**Comment by MiguelCastillo** [comment]
@kesne it would be really wonderful if we could get some traction on this PR :)

**Comment by marijnh** [comment]
(As this is one of the more popular Acorn plugins, it would be really great if we could get a 6.0-compatible release soon!)

File: package.json [hunk 1]
```diff
@@ -24,10 +24,13 @@
     "preversion": "npm run test && npm run check-changelog && npm run check-only-changelog-changed",
     "tag": "git tag v$npm_package_version"
   },
-  "dependencies": {
-    "acorn": "^5.0.0"
+  "peerDependencies": {
+    "acorn": "^6.0.0",
+    "acorn-walk": "^6.0.0"
```
Comment (kesne) [inline package.json]: It feels weird to have this be a peer depencency, but have this functionality be optional. Can this either be a regular dependency, or can we just make a note in the readme that if you're going to use the walk plugin you need to install this?

**Comment by kesne** [comment]
Sorry, I don't read my github notifications often enough so I missed this. I think this should be good to go if we remove the acorn-walk peer dependency.

File: package.json (hunk 1 above)
Comment (marijnh) [reply package.json]: Would dropping the re-export from `walk.js` (along with the peer dep) be acceptable to you? That way, it's a little more obvious that you'll need to install acorn-walk yourself.

Comment (marijnh) [reply package.json]: As proposed in 9841d70

Comment (kesne) [reply package.json]: Yeah I think that's reasonable.

**Comment by adrianheine** [comment]
Thanks! I can update acorn-stage3 when you do a release.

**Comment by kesne** [comment]
@adrianheine Released as 4.0.0.

**Comment by adrianheine** [comment]
Nice, I released acorn-stage3 0.7.0.

**Comment by marijnh** [comment]
Thanks!

**Pull request merged**